src/lib/coda_metrics/coda_metrics.ml

```diff
@@ -90,6 +90,43 @@ module TextFormat_0_0_4 = struct
           samples )
 end
 
+module Moving_average (Spec : sig
+  val tick_interval : Core.Time.Span.t
+
+  val rolling_interval : Core.Time.Span.t
+
+  val display : Float.t -> Float.t
+end) =
+struct
+  open Async
+
+  let data_set = ref []
+
+  let create () ~name ~subsystem ~namespace ~help =
+    let gauge = Gauge.v name ~subsystem ~namespace ~help in
+    let rec tick () =
+      upon (after Spec.tick_interval) (fun () ->
+          let current_time = Core.Time.now () in
+          data_set :=
+            List.filter !data_set ~f:(fun (timestamp, _) ->
+                let open Core.Time in
+                sub current_time Spec.rolling_interval < timestamp ) ;
+          let total =
+            List.fold !data_set ~init:0. ~f:(fun accm (_, datum) ->
+                accm +. datum )
```

Comment:
This seems really inefficient. Wouldn't it be better to use a query instead of a list to store the data and the time stamps. It's guaranteed that the tuples in the queue will be ordered from oldest to latest from the head of the queue to the tail of the queue. So, whenever a timestamp expires, you can just pop the head of the queue. Also, when you want to add a new datapoint, you add it to the tail of the queue.

On top of having a queue, you can have an extra memoized `float` value that represents the aggregate value that you want to compute. I think this aggregate value is just the sum of all values in the queue.